exonum-java-binding-core/rust/src/utils/jni_cache.rs

```diff
@@ -0,0 +1,158 @@
+
+#[allow(non_snake_case)]
+#[no_mangle]
+pub extern "system" fn JNI_OnLoad(vm: JavaVM, _: *mut c_void) -> jint {
+    let env = vm.get_env().expect("Cannot get reference to the JNIEnv");
+    cache_methods(&env);
+
+    JNI_VERSION_1_8
+}
+
+/// Caches all required classes and methods ids.
+pub fn cache_methods(env: &JNIEnv) {
+    unsafe {
+        OBJECT_CACHED_GET_CLASS =
+            get_method_id(&env, "java/lang/Object", "getClass", "()Ljava/lang/Class;");
+        CLASS_CACHED_GET_NAME =
+            get_method_id(&env, "java/lang/Class", "getName", "()Ljava/lang/String;");
+        THROWABLE_CACHED_GET_MESSAGE = get_method_id(
+            &env,
+            "java/lang/Throwable",
+            "getMessage",
+            "()Ljava/lang/String;",
+        );
+        UTA_CACHED_EXECUTE = get_method_id(
+            &env,
+            "com/exonum/binding/service/adapters/UserTransactionAdapter",
+            "execute",
+            "(J)V",
+        );
+        UTA_CACHED_INFO = get_method_id(
+            &env,
+            "com/exonum/binding/service/adapters/UserTransactionAdapter",
+            "info",
+            "()Ljava/lang/String;",
+        );
+        UTA_CACHED_VERIFY = get_method_id(
+            &env,
+            "com/exonum/binding/service/adapters/UserTransactionAdapter",
+            "isValid",
+            "()Z",
+        );
+        USA_CACHED_STATE_HASHES = get_method_id(
+            &env,
+            "com/exonum/binding/service/adapters/UserServiceAdapter",
+            "getStateHashes",
+            "(J)[[B",
+        );
+        USA_CACHED_CONVERT_TRANSACTION = get_method_id(
+            &env,
+            "com/exonum/binding/service/adapters/UserServiceAdapter",
+            "convertTransaction",
+            "([B)Lcom/exonum/binding/service/adapters/UserTransactionAdapter;",
+        );
+        CLASS_CACHED_ERROR = env
+            .new_global_ref(env.find_class("java/lang/Error").unwrap().into())
+            .ok();
+        CLASS_CACHED_TEE = env
+            .new_global_ref(
+                env.find_class("com/exonum/binding/transaction/TransactionExecutionException")
+                    .unwrap()
+                    .into(),
+            ).ok();
+
+        assert!(
+            OBJECT_CACHED_GET_CLASS.is_some()
+                && CLASS_CACHED_ERROR.is_some()
+                && THROWABLE_CACHED_GET_MESSAGE.is_some()
+                && UTA_CACHED_EXECUTE.is_some()
+                && UTA_CACHED_INFO.is_some()
+                && UTA_CACHED_VERIFY.is_some()
+                && USA_CACHED_STATE_HASHES.is_some()
+                && USA_CACHED_CONVERT_TRANSACTION.is_some()
+                && CLASS_CACHED_ERROR.is_some()
+                && CLASS_CACHED_TEE.is_some(),
+            "Error caching Java entities"
+        );
+
+        info!("Done caching references to Java classes and methods.");
+    }
+}
+
+// Helper method. Produces JMethodID for a particular class dealing with lifetime.
```

Comment:
It should be doc comment started with `///`

Reply:
Also I encourage to use backticks around every type in docstring and comments - `JMethodID`, not JMethodID

Reply:
Yeap I agree about backticks (just missed them in this case), but about doc comments - don't we do this for public methods only, do we?

Reply:
Doc comments are invisible for private methods, but they must be supported by IDE, for example

Reply:
Well, looking at our code we didn't follow this rule before, but if the general approach is like "not required, but nice to have" then it's better to add a spare slash ;-)

Reply:
For methods that do not have `public` visibility — use your judgement to decide if they need a documentation based on their frequency of use, complexity, their name, distance from their usage, etc. I would agree, however, that if you do add a comment to the method, make it a doc comment so that it can be easily looked up in an IDE (and gets appropriate highlighting). If anything isn't this way, it's a bug.